triton-vm/src/table/master_table.rs

```diff
@@ -1046,59 +1048,117 @@ pub fn all_quotients(
-        consistency_quotient_zerofier_inverse(trace_domain, quotient_domain).view(),
-        challenges,
-    );
-    prof_stop!(maybe_profiler, "consistency");
-
-    prof_start!(maybe_profiler, "transition", "AIR");
-    MasterExtTable::fill_transition_quotients(
-        quotient_domain_master_base_table,
-        quotient_domain_master_ext_table,
-        &mut quotient_table.slice_mut(s![.., cons_section_end..tran_section_end]),
-        transition_quotient_zerofier_inverse(trace_domain, quotient_domain).view(),
-        challenges,
-        trace_domain,
-        quotient_domain,
+
+    prof_start!(maybe_profiler, "zerofier inverse");
+    let initial_zerofier_inverse = initial_quotient_zerofier_inverse(quotient_domain);
+    let consistency_zerofier_inverse =
+        consistency_quotient_zerofier_inverse(trace_domain, quotient_domain);
+    let transition_zerofier_inverse =
+        transition_quotient_zerofier_inverse(trace_domain, quotient_domain);
+    let terminal_zerofier_inverse =
+        terminal_quotient_zerofier_inverse(trace_domain, quotient_domain);
+    prof_stop!(maybe_profiler, "zerofier inverse");
+
+    prof_start!(
+        maybe_profiler,
+        "evaluate AIR / divide zerofiers / combine quotients"
     );
-    prof_stop!(maybe_profiler, "transition");
-
-    prof_start!(maybe_profiler, "terminal", "AIR");
-    MasterExtTable::fill_terminal_quotients(
-        quotient_domain_master_base_table,
-        quotient_domain_master_ext_table,
-        &mut quotient_table.slice_mut(s![.., tran_section_end..term_section_end]),
-        terminal_quotient_zerofier_inverse(trace_domain, quotient_domain).view(),
-        challenges,
+    let quotient_codeword = (0..quotient_domain.length)
+        .into_par_iter()
+        .map(|row| {
+            // get views of relevant rows
+            let unit_distance = quotient_domain.length / trace_domain.length;
+            let domain_length_bit_mask = quotient_domain.length - 1;
+            let next_row_index = (row + unit_distance) & domain_length_bit_mask;
+            let current_row_main = quotient_domain_master_base_table.slice(s![row, ..]);
+            let current_row_aux = quotient_domain_master_ext_table.slice(s![row, ..]);
+            let next_row_main = quotient_domain_master_base_table.slice(s![next_row_index, ..]);
+            let next_row_aux = quotient_domain_master_ext_table.slice(s![next_row_index, ..]);
+
+            // initial constraints
+            let initial_constraint_values = MasterExtTable::evaluate_initial_constraints(
+                current_row_main,
+                current_row_aux,
+                challenges,
+            );
+            let initial_inner_product = initial_constraint_values
+                .into_iter()
+                .zip_eq(
+                    quotient_weights
+                        .iter()
+                        .take(MasterExtTable::NUM_INITIAL_CONSTRAINTS),
+                )
```

Comment:
```suggestion
                .zip_eq(quotient_weights[..init_section_end].iter())
```

Reply:
Same for the other three; see the conversations marked as “resolved” for copy-pasteable code.